### PR TITLE
Add constant time code for BN_is_bit_set call.

### PR DESCRIPTION
**Pull request by paulidale**
The `BN_is_bit_set` call isn't constant time, specifically when the bit asked for is larger than the topmost bit of the final limb.  This change makes that case constant time.

The changed code path only applies to big numbers that have the constant time flag set.

Reported by Samuel Weiser.

- [x] tests are added or updated

### CODE REVIEW

**Comment by davidben** [comment]
This isn't constant-time either. The same information is leaked via memory access pattern.

Some code does play games with `a->dmax` vs `a->top`, which addresses this sort of thing in some scenarios, but at some point one really should just fix the `bn_correct_top` issue and avoid adding workaround atop workaround. (All of this stuff is unnecessary in a world where secret `BIGNUM`s maintain their public bit widths correctly.)

Additionally, conditioning on `BN_FLG_CONSTTIME` like that is an anti-pattern per #6078.

**Comment by paulidale** [comment]
I realised the memory access pattern leaked something.
I've switched to a read everything approach instead which could leak the length.  Maybe dmax is the way to go.

I do agree that changing to fixed length private BNs is sensible and the way forward -- that a step we're a way from making (unfortunately).

**Comment by paulidale** [comment]
Change to iterate over all words up to _dmax_ in the constant time case.

**Comment by davidben** [comment]
Iterating over all the words is if the bit you're looking up is secret. I don't know of any uses where it is. Here, the bit is public, but you are attempting to keep `bn->top` secret. That means you should:
* If the index is beyond `dmax`, there is no hope and you just early return.
* Otherwise, read the element directly and mask off the value if beyond `top`

Fixing the `bn_correct_top` issue is something that can be worked on incrementally. I laid out one possible approach in #6640. It's work, yes, but it is quite tractable. I did it for BoringSSL in a few weeks.

**Comment by paulidale** [comment]
Third time lucky, a fast bounds check against the allocated size of the BN followed by a constant time access and filter if over the top.

**Comment by kroeckx** [comment]
Is using something like BN_num_bits() an alternative to BN_is_bit_set()?

**Comment by paulidale** [comment]
It would work in this case but wouldn't be general.
I think we're better moving towards a general solution to the constant time issues.

The iteration over the dmax limbs would also serve in this case -- we're interested in the bit (possibly) one past the top.  David's suggestion is better all around.

**Comment by kroeckx** [comment]
Yes, iterating over the all allocated limbs was my proposal. But I think it's also important to document which side channels it protects against. We need to know which data we want to protect, and then use a function that is documented to protect against that.

I think we should also document the callers of those functions to say what we expect to be protecting.

**Comment by davidben** [comment]
(This is still conditioning on `BN_FLG_CONSTTIME`, an anti-pattern per #6078.)

Definitely do not iterate over all the limbs. That would be a performance regression to much of the library. It is very common to extract a the nth bit of a number, where n is public, but the number is not. It's rare to have n be secret. The cost of making n secret is going from an O(1) operation to O(N), so they should be separate functions.

However, even the natural documentation of "n is public and the number is secret" does not cover this function, because this PR breaks `BIGNUM`'s abstractions. It is purely a workaround for a deeper design problem in OpenSSL.

You cannot have a secret number without a public upper bound, for the same reason that you cannot have a secret byte array without a public length. Every time a function's contract is "foo is treated as secret", there is an implicit bound. `CRYPTO_memcmp`'s length is public. Secret AES keys have a public size (128, 192, or 256 bits). The various constant-time helper functions used fixed-width integers which are implicitly bounded.

However, `BIGNUM`'s abstraction doesn't support a public bound. What bound do we want? Well, we often work with numbers mod n, where n is public, so the public bound should be n or something rounded up (bit width of n?). Sometimes we work with numbers mod p where p is secret (RSA CRT), so the public bound is even looser. (For RSA, p's public bound is 2^{bit width} since p's bit width is half of n's bit width. This gets a little fuzzy with multi-prime RSA, but it'll do.)

This public bound must be understood in `BIGNUM` itself because it controls how you represent the value in memory. Zero with a public bound of 2^256 must be represented in a way that also supports 2^256-1 with a public bound of 2^256. Namely we must zero-pad the array accordingly.

Since `BIGNUM`s are ultimately arrays, that gives a natural public upper bound: the bounds on `bn->d`, namely `bn->top`. Indeed that is how most constant-time code in OpenSSL works. Note how `BN_mod_exp_mont_consttime` uses `p->top` to determine the number of bits in the exponent.

Under that interpretation, `BN_is_bit_set` already respected the "n is public, number is secret" contract. Indeed that contract is satisfiable without performance hit, so there's no need for a separate function or `BN_FLG_CONSTTIME`...

Except, there's still a leak. Why? `bn->top` is wrong. It is the natural public upper bound, but the `bn_correct_top` invariant is unsuitable. So we apply workarounds. `bn->top` is secret too, but maybe we got lucky and `bn->dmax` is right, though the abstraction doesn't favor it (need to mask off values because they may not be zero), `bn->dmax` is an implementation detail, and `bn->top` dependence is ubiquitous.

But these are just workarounds, not fixes. Where the timing variation is large or clearly exploitable, a workaround to reduce it to a small, maybe(?) unexploitable timing variation makes sense. When the timing variation is small, you won't eliminate it anyway. To fix small timing variations, we need to fix `bn->top`, at which point all these workarounds can be removed.

It turns out fixing `bn->top` is quite tractable. It makes sense to just do it at this point.

File: crypto/bn/bn_lib.c [hunk 1]
```diff
@@ -649,15 +649,27 @@ int BN_clear_bit(BIGNUM *a, int n)
...
-    if (a->top <= i)
-        return 0;
-    return (int)(((a->d[i]) >> j) & ((BN_ULONG)1));
+
+    if (a->flags & BN_FLG_CONSTTIME) {
+        int it;
+
+        if (i >= a->dmax)
+            return 0;
+        it = constant_time_lt(i, a->top);
```
Comment (kroeckx) [inline crypto/bn/bn_lib.c]: I think this should use a->dmax, not a->top. As far as I know everything between top and dmax should be set to 0.

Comment (kroeckx) [reply crypto/bn/bn_lib.c]: And so this line should just get removed ...

**Comment by kroeckx** [comment]
So my current understanding of what needs to happen in this function is:
- Check dmax instead of top
- Document that n is considered public information, but that bignum can be a secret value.

**Comment by kroeckx** [comment]
That documentation change should mention that you need to make sure that the size is always the same, but we currently don't actually seem to have a public function to do that. bn_expand() is internal only.

**Comment by kroeckx** [comment]
But as @davidben explains, it might actually not be enough, and we might need to consider fixing whole bignum instead.

File: crypto/bn/bn_lib.c (hunk 1 above)
Comment (davidben) [reply crypto/bn/bn_lib.c]: Sadly, no, stuff between `top` and `dmax` is not set to zero. For instance, this function:
https://github.com/openssl/openssl/blob/master/crypto/bn/bn_lib.c#L854-L859

(This is why I'm not a huge fan of playing `top` vs `dmax` games.)

Comment (paulidale) [reply crypto/bn/bn_lib.c]: It has to use top, as @davidben wrote, there is no guarantee of zeroes between top and gmax.  The code has to behave as if there are zeroes there but still read what's really there...

**Comment by paulidale** [comment]
Closing this one.  It seems like properly fixing BN is the way to go instead of applying band aids.

**Pull request closed without merging**